pkg/fqdn/dnsproxy/proxy.go

```diff
@@ -218,50 +218,87 @@ func (p *DNSProxy) checkRestored(endpointID uint64, destPort uint16, destIP stri
+func (p *DNSProxy) skipIPInRestoration(ip string) bool {
+	p.RLock()
+	defer p.RUnlock()
+	if len(p.usedServers) > 0 {
+		if _, used := p.usedServers[ip]; !used {
+			return true
+		}
+	}
+	return false
+}
+
 // GetRules creates a fresh copy of EP's DNS rules to be stored
 // for later restoration.
 func (p *DNSProxy) GetRules(endpointID uint16) (restore.DNSRules, error) {
+	// Lock ordering note: Acquiring the IPCache read lock (as LookupIPsBySecID does) while holding
+	// the proxy lock can lead to a deadlock. Avoid this by reading the state from DNSProxy while
+	// holding the read lock, then perform the IPCache lookups.
+	// Note that IPCache state may change in between calls to LookupIPsBySecID.
 	p.RLock()
-	defer p.RUnlock()
 
-	restored := make(restore.DNSRules)
+	type selRegex struct {
+		re *regexp.Regexp
+		cs policy.CachedSelector
+	}
+
+	portToSelRegex := make(map[uint16][]selRegex)
 	for port, entries := range p.allowed[uint64(endpointID)] {
-		var ipRules restore.IPRules
+		var nidRules = make([]selRegex, 0, len(entries))
+		// Copy the entries to avoid racy map accesses after we release the lock. We don't need
+		// constant time access, hence a preallocated slice instead of another map.
 		for cs, regex := range entries {
-			var IPs map[string]struct{}
-			if !cs.IsWildcard() {
-				IPs = make(map[string]struct{})
-				count := 0
-			Loop:
-				for _, nid := range cs.GetSelections() {
-					nidIPs := p.LookupIPsBySecID(nid)
-					for _, ip := range nidIPs {
-						// Skip IPs that are allowed but have never been used,
-						// but only if at least one server has been used so far.
-						if len(p.usedServers) > 0 {
-							if _, used := p.usedServers[ip]; !used {
-								continue
-							}
-						}
-						IPs[ip] = struct{}{}
-						count++
-						if count > p.maxIPsPerRestoredDNSRule {
-							log.WithFields(logrus.Fields{
-								logfields.EndpointID:            endpointID,
-								logfields.Port:                  port,
-								logfields.EndpointLabelSelector: cs,
-								logfields.Limit:                 p.maxIPsPerRestoredDNSRule,
-								logfields.Count:                 len(nidIPs),
-							}).Warning("Too many IPs for a DNS rule, skipping the rest")
-							break Loop
-						}
+			nidRules = append(nidRules, selRegex{cs: cs, re: regex})
+		}
+		portToSelRegex[port] = nidRules
+	}
+
+	maxIPsPerRestoredDNSRule := p.maxIPsPerRestoredDNSRule
+
+	// We've read what we need from the proxy. The following IPCache lookups _must_ occur outside of
+	// the critical section.
+	p.RUnlock()
+
+	restored := make(restore.DNSRules)
+	for port, selRegexes := range portToSelRegex {
+		var ipRules restore.IPRules
+		for _, selRegex := range selRegexes {
+			if selRegex.cs.IsWildcard() {
+				ipRules = append(ipRules, asIPRule(selRegex.re, nil))
+				continue
+			}
+			ips := make(map[string]struct{})
+			count := 0
+			nids := selRegex.cs.GetSelections()
+		Loop:
+			for _, nid := range nids {
+				nidIPs := p.LookupIPsBySecID(nid)
```

Comment:
It occurs to me that it's this callback mechanism that's injected in the `StartDNSProxy()` function that never really stated any sort of locking constraints... probably we should at least document if not rework the way that lookup occurs at some point to make it less error-prone. Maybe a target for future `pkg/statedb` work.

This raises an interesting question though because `LookupIPsBySecIDFunc` is just one of three callbacks there that we may want to investigate if we haven't already, and confirm the locking behaviours for each of them.

Reply:
Yes, I did check the other callbacks and did not find anything. They mostly seem to be called when a DNS request is handled, which does not hold the EP or DNSProxy lock (or only for a short, isolated time).